### PR TITLE
Ats magick

### PR DESCRIPTION
**Pull request by dmorilha**
new ats-magick transformation plug-in.

ImageMagick dependencies have to be added into the build environment to be able to compile this. Not sure on how to handle this.

### CODE REVIEW

**Comment by bryancall** [comment]
experimental/magick/magick.cc:42:10: fatal error: 'MagickWand/MagickWand.h' file not found
#include <MagickWand/MagickWand.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~

We should detect if imagemagick is installed and only build the plugin if it is installed.

File: plugins/experimental/magick/test.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#/bin/sh
+set -e -x -u;
+DIR=${DIR:=`mktemp -d;`}
+convert netlify.png $@ $DIR/netlify-2.png;
```
Comment (maskit) [inline plugins/experimental/magick/test.sh]: We should probably use trafficserver's logo instead.

Comment (dmorilha) [reply plugins/experimental/magick/test.sh]: The old or the new? I heard there is a contest going on...

File: plugins/experimental/magick/magick.cc [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+
+#include <MagickWand/MagickWand.h>
+#include <MagickWand/magick-cli.h>
+
+#define PLUGIN_TAG "ats-magick"
+
+using std::string;
+using namespace atscppapi;
+
+typedef std::vector<char> CharVector;
```
Comment (maskit) [inline plugins/experimental/magick/magick.cc]: IIRC, we switched to `using` type alias.

File: plugins/experimental/magick/test.sh (hunk 1 above)
Comment (maskit) [reply plugins/experimental/magick/test.sh]: As far as I know there's no new logo yet. Current logo is in doc/static/images/press .

**Comment by dmorilha** [comment]
All code review comments were addressed. Do we have an ETA for the merge? I would like to write a blog post here (https://www.netlify.com/blog/) and link back to github before Friday. Thank you folks for reviewing it.

**Comment by maskit** [comment]
To make RAT test happy:
- Put Apache License header to shell scripts
- Add ".key" suffix to the key filenames

It seems like some code could be simplify with C++17 (e.g. Iteration for std::vector) but it's not a big deal. As we talked on the last ATS Summit, it would be nice if this plugin uses a dedicated thread for image processing to not block event threads. For now, it might be better to put some note about it.

Looks good overall. Please squash the commits into one commit after fixing RAT test errors.

**Comment by shukitchan** [comment]
as I mentioned in the summit, I think we should combine the webp transform plugin with this one.  And you probably need to take care of the content type to support that. For now it is fine to be experimental. Perhaps you can do this before promoting it to stable plugin.

File: plugins/experimental/magick/magick.cc [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+      TSDebug(PLUGIN_TAG, "Destroying thread number %lu (%p)", i, thread);
+      TSThreadWait(thread);
+      TSThreadDestroy(thread);
+    }
+
+    delete[] pool_;
+    const_cast<TSThread *&>(pool_) = nullptr;
+  }
+
+  ThreadPool(const size_t s) : size_(s), running_(true), pool_(new TSThread[s])
```
Comment (dmorilha) [inline plugins/experimental/magick/magick.cc]: thread-pool

**Comment by dmorilha** [comment]
@shukitchan not opposed to that idea at all in the mid-term, but for now let's just land it if we could. Thanks.

File: plugins/experimental/magick/magick.cc (hunk 3 above)
Comment (maskit) [reply plugins/experimental/magick/magick.cc]: Nice!

**Comment by maskit** [comment]
[approve ci autest]

**Pull request merged**